### PR TITLE
feat: support store secret in env for auth plugin

### PR DESCRIPTION
**Pull request by soulbird**
### Description

As part of https://github.com/apache/apisix/issues/8319:

Step 1: Support the use of environment variables to manage keys. Information in the consumer configuration is more private, so we first used environment variables in the consumer to manage sensitive information.

Currently only test cases have been added for the key-auth plugin, the test cases for other plugins will be completed in a separate PR.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/core/env.lua [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  local e = ffi.C.environ
+  if not e then
+    log.warn("could not access environment variables")
+    return
+  end
+
+  local i = 0
+  while e[i] ~= nil do
+    local var = str(e[i])
+    local p = find(var, "=")
```
Comment (spacewander) [inline apisix/core/env.lua:50]: We can use core.string.find, which is a plaintext find.

File: apisix/core/env.lua [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+function _M.get(ref)
+    if not is_env_ref(ref) then
+        return nil
+    end
+
+    local key, sub_key = parse_ref(ref)
+    local main_value = apisix_env_vars[key] or os.getenv(key)
+    if main_value and sub_key ~= "" then
+        local vt = json.decode(main_value)
+        if not vt then
```
Comment (spacewander) [inline apisix/core/env.lua:87]: Let's log down the error when the decode is failed.

File: apisix/core/utils.lua [hunk 3]
```diff
@@ -329,4 +332,57 @@ end
...
+                retrieve_ref(v)
+            end
+        end
+        return refs
+    end
+
+    local function retrieve(refs)
+        log.info(string.format("retrieve secrets refs: %p", refs))
+
+        local new_refs = table.clone(refs)
```
Comment (spacewander) [inline apisix/core/utils.lua:357]: Do we need to use deepcopy for the later recursive call?

Comment (soulbird) [reply apisix/core/utils.lua]: Actually I'm not sure, but it seems safer to use deepcopy.

**Review by spacewander — DISMISSED** [review]

File: apisix/core/env.lua [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  local e = ffi.C.environ
+  if not e then
+    log.warn("could not access environment variables")
+    return
+  end
+
+  local i = 0
+  while e[i] ~= nil do
+    local var = str(e[i])
+    local p = find(var, "=")
```
Comment (membphis) [inline apisix/core/env.lua:51]: we should use `plain` model, all right?

Comment (membphis) [reply apisix/core/env.lua:51]: which is better performance

File: apisix/core/env.lua [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        apisix_env_vars[sub(var, 1, p - 1)] = sub(var, p + 1)
+    end
+
+    i = i + 1
+  end
+end
+
+
+local function is_env_ref(ref)
+    return type(ref) == "string" and sub(upper(ref), 1, 7) == ENV_PREFIX
```
Comment (membphis) [inline apisix/core/env.lua:62]: we can use `C.memcmp` which is better performance

File: apisix/core/env.lua [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+end
+
+
+local function parse_ref(ref)
+    local path = sub(ref, 8)
+    local idx = find(path, "/")
```
Comment (membphis) [inline apisix/core/env.lua:68]: use `plain` model

File: apisix/core/env.lua [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    return key, sub_key
+end
+
+
+function _M.get(ref)
+    if not is_env_ref(ref) then
+        return nil
+    end
+
+    local key, sub_key = parse_ref(ref)
```
Comment (membphis) [inline apisix/core/env.lua:84]: I think we should use `array` object

File: apisix/core/utils.lua [hunk 8]
```diff
@@ -329,4 +332,57 @@ end
...
+                refs[k] = env.get(v) or v
+            elseif typ == "table" then
+                retrieve_ref(v)
+            end
+        end
+        return refs
+    end
+
+    local function retrieve(refs)
+        log.info(string.format("retrieve secrets refs: %p", refs))
```
Comment (membphis) [inline apisix/core/utils.lua:355]: is it maybe in `hot` code path? if yes, we need to cache `string.format`

File: apisix/core/env.lua (hunk 4 above)
Comment (soulbird) [reply apisix/core/env.lua:51]: Already use `local string = require("apisix.core.string")`, it's `plain` mode.

File: apisix/core/env.lua (hunk 6 above)
Comment (soulbird) [reply apisix/core/env.lua:68]: ditto

File: apisix/core/env.lua (hunk 4 above)
Comment (membphis) [reply apisix/core/env.lua:51]: @soulbird you are right, ignore my comments

File: apisix/core/utils.lua (hunk 8 above)
Comment (soulbird) [reply apisix/core/utils.lua:355]: Not a hot path, but it would be better to use cache.

File: apisix/core/env.lua [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+end
+
+
+function _M.get(ref)
+    if not is_env_ref(ref) then
+        return nil
+    end
+
+    local opts = parse_ref(ref)
+    local main_value = apisix_env_vars[opts.key] or os.getenv(opts.key)
```
Comment (tokers) [inline apisix/core/env.lua:93]: Why still try `os.getenv` here?

File: apisix/core/env.lua [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    if main_value and opts.sub_key ~= "" then
```
Comment (tokers) [inline apisix/core/env.lua:94]: Where is the convention about the value format (a JSON string), missing docs will make users confusing.

Comment (tokers) [reply apisix/core/env.lua:94]: Also, looks like we only support one level index, that also needs to be written.

File: apisix/core/env.lua [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    i = i + 1
+  end
+end
+
+
+local function is_env_ref(ref)
+    -- We will not use string.has_prefix,
+    -- to avoid the error caused by has_prefix to cause a crash.
+    return type(ref) == "string" and #ref > 7 and
+        0 == C.memcmp(upper(ref), ENV_PREFIX, 7)
```
Comment (spacewander) [inline apisix/core/env.lua:67]: `type(ref) == "string" and string.has_prefix(upper(ref), ENV_PREFIX)`, will this way cause a crash?

File: apisix/core/env.lua [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+local apisix_env_vars = {}
+
+ffi.cdef [[
+  extern char **environ;
+  int memcmp(const void *s1, const void *s2, size_t n);
+]]
+
+
+function _M.init()
+  local e = ffi.C.environ
```
Comment (spacewander) [inline apisix/core/env.lua:44]:
```suggestion
  local e = C.environ
```

File: apisix/core/env.lua [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+local function is_env_ref(ref)
+    -- We will not use string.has_prefix,
+    -- to avoid the error caused by has_prefix to cause a crash.
+    return type(ref) == "string" and #ref > 7 and
+        0 == C.memcmp(upper(ref), ENV_PREFIX, 7)
+end
+
+
+local function parse_ref(ref)
+    local path = sub(ref, 8)
```
Comment (spacewander) [inline apisix/core/env.lua:72]:
```suggestion
    local path = sub(ref, #ENV_PREFIX + 1)
```
can remove the magic number.

File: t/core/env.t [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        }
+    }
+--- request
+GET /t
+--- response_body
+nil
+
+
+
+=== TEST 4: worong format: worong type
```
Comment (tokers) [inline t/core/env.t:80]:
```suggestion
=== TEST 4: wrong format: wrong type
```

File: apisix/core/env.lua (hunk 11 above)
Comment (soulbird) [reply apisix/core/env.lua:67]: No, but `type(ref) == "string"` will be executed one more time.

File: apisix/core/env.lua (hunk 9 above)
Comment (soulbird) [reply apisix/core/env.lua:93]: In order to be compatible with `env foo=bar;` this Nginx configuration.

File: apisix/core/env.lua (hunk 10 above)
Comment (soulbird) [reply apisix/core/env.lua:94]: The docs will be completed at the end, I'll notice you mention this.

Comment (tokers) [reply apisix/core/env.lua:94]: Got it.

File: apisix/core/env.lua (hunk 9 above)
Comment (tokers) [reply apisix/core/env.lua:93]: Then why now also handle them when init the `apisix_env_vars`.

File: apisix/core/env.lua (hunk 11 above)
Comment (spacewander) [reply apisix/core/env.lua:67]: IMHO, check type twice is better than reinventing a function. The type check is cheap.

Comment (soulbird) [reply apisix/core/env.lua:67]: OK

File: apisix/core/env.lua (hunk 9 above)
Comment (soulbird) [reply apisix/core/env.lua:89]: Means `Then why not also`?
Because I didn't find a way to iterate through them, only by a specific name.

**Review by spacewander — APPROVED** [review]

File: apisix/core/env.lua (hunk 9 above)
Comment (tokers) [reply apisix/core/env.lua:89]: ok

File: t/core/env.t (hunk 14 above)
Comment (tokers) [reply t/core/env.t:80]: Forget to change this?

**Review by tokers — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Review by spacewander — APPROVED** [review]

**Pull request merged**